File: exercises/posix-thead/exercise02.cpp
// =================================================================
//
// File: exercise02.cpp
// Author(s):
// Description: This file contains the code that performs the sum of
//				all prime numbers less than or equal to MAXIMUM using
//				pthreads.
//              To compile: g++ exercise02.cpp -lpthread
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include "utils.h"
#include <algorithm>
#include <cmath>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <omp.h>
#include <pthread.h>

#define MAXIMUM 1000000 // 1e6

using namespace std;

const int THREADS = 4;

typedef struct
{
	int start, end; // [start, end)
} Block;

void *sumPrimes(void *param)
{
	int *acum;
	Block *block;
	int i, j, limit;

	block = (Block *)param;
	acum = new int;
	(*acum) = 0;
	for (i = block->start; i < block->end; i++)
	{
		limit = (int)sqrt(i);
		for (j = 2; j <= limit; j++)
		{
			if (i % j == 0)
			{
				break;
			}
		}
		if (j > limit)
		{
			(*acum) += i;
		}
	}
	return ((void **)acum);
}

int main(int argc, char *argv[])
{
	int block_size, i, j;
	double ms, result, *acum;
	Block blocks[THREADS];
	pthread_t tids[THREADS];

	cout << "Starting..." << endl;
	ms = 0;

	block_size = MAXIMUM / THREADS;

	for (i = 0; i < THREADS; i++)
	{
		blocks[i].start = i * block_size;
		if (i != (THREADS - 1))
		{
			blocks[i].end = blocks[i].start + block_size;
		}
		else
		{
			blocks[i].end = MAXIMUM;
		}
	}

	for (int i = 0; i < NUM; i++)
	{
		start_timer();

		result = 0;
		for (j = 0; j < THREADS; j++)
		{
			pthread_create(&tids[j], NULL, sumPrimes, (void *)&blocks[j]);
		}

		ms += stop_timer();
	}
	cout << "result = " << fixed << setprecision(0) << result << "\n";
	cout << "avg time = " << fixed << setprecision(5) << (ms / NUM) << " ms" << endl;

	return 0;
}

File: exercises/processes-python/exercise02.py
# =================================================================
##
# File: exercise02.py
# Author(s): Esteban Padilla Cerdio - A01703068
# 			 Hilda Olivia Beltrán Acosta - A01251916
# Description: This file contains the code that performs the sum of
# all prime numbers less than or equal to MAXIMUM
# using processes in Python.
##
# Copyright (c) 2022 by Tecnologico de Monterrey.
# All Rights Reserved. May be reproduced for any non-commercial
# purpose.
##
# =================================================================

import utils
import time
import multiprocessing as mp
import threading

MAXIMUM = 1000000  # 1e6


def isPrime(n):
    if n <= 1:
        return False
    elif n <= 3:
        return True
    elif n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True


def partialSumPrime(start, end, queue):
    sum = 0
    for i in range(start, end):
        if isPrime(i):
            sum += i
    queue.put(sum)


if __name__ == "__main__":

    print("Starting...")
    startTime = endTime = ms = 0

    CORES = 10
    blockSize = MAXIMUM // CORES

    for i in range(utils.N):

        startTime = time.time() * 1000

        queue = mp.SimpleQueue()
        processes = list()

        for i in range(CORES):
            start = i * blockSize

            if i != (CORES - 1):
                end = (i + 1) * blockSize
            else:
                end = MAXIMUM
            process = mp.Process(target=partialSumPrime,
                                 args=(start, end, queue,))
            processes.append(process)
            process.start()

        result = 0

        for i in range(CORES):
            result += queue.get()

        endTime = time.time() * 1000

        ms = ms + (endTime - startTime)

    print("sum = ", result)
    print("avg time = ", (ms / utils.N), " ms")
